From 1e302447d2e1197f49622a33b6edaf636059e772 Mon Sep 17 00:00:00 2001
From: Doug Ledford <dledford@redhat.com>
Date: Wed, 24 Sep 2014 17:27:55 -0400
Subject: [infiniband] ehca: Returns an error on ib_copy_to_udata() failure

Message-id: <7bb799fa35be16b576de9711c9792baf2143298f.1411578494.git.dledford@redhat.com>
Patchwork-id: 96663
O-Subject: [Patch RHEL7 02/26] IB/ehca: Returns an error on ib_copy_to_udata() failure
Bugzilla: 1107621
RH-Acked-by: Jes Sorensen <Jes.Sorensen@redhat.com>
RH-Acked-by: Steve Dickson <SteveD@redhat.com>

From: Yann Droneaud <ydroneaud@opteya.com>

In case of error when writing to userspace, function ehca_create_cq()
does not set an error code before following its error path.

This patch sets the error code to -EFAULT when ib_copy_to_udata()
fails.

This was caught when using spatch (aka. coccinelle)
to rewrite call to ib_copy_{from,to}_udata().

Link: https://www.gitorious.org/opteya/coccib/source/75ebf2c1033c64c1d81df13e4ae44ee99c989eba:ib_copy_udata.cocci
Link: http://marc.info/?i=cover.1394485254.git.ydroneaud@opteya.com
Cc: <stable@vger.kernel.org>
Signed-off-by: Yann Droneaud <ydroneaud@opteya.com>
Signed-off-by: Roland Dreier <roland@purestorage.com>
(cherry picked from commit 5bdb0f02add5994b0bc17494f4726925ca5d6ba1)
Signed-off-by: Doug Ledford <dledford@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/infiniband/hw/ehca/ehca_cq.c b/drivers/infiniband/hw/ehca/ehca_cq.c
index 212150c..8cc8375 100644
--- a/drivers/infiniband/hw/ehca/ehca_cq.c
+++ b/drivers/infiniband/hw/ehca/ehca_cq.c
@@ -283,6 +283,7 @@ struct ib_cq *ehca_create_cq(struct ib_device *device, int cqe, int comp_vector,
    (my_cq->galpas.user.fw_handle & (PAGE_SIZE - 1));
   if (ib_copy_to_udata(udata, &resp, sizeof(resp))) {
    ehca_err(device, "Copy to udata failed.");
+   cq = ERR_PTR(-EFAULT);
    goto create_cq_exit4;
   }
  }
-- 
1.7.1